Add metrics for config reloads and config hash

jcreixell opened the pull request:
#### PR Description

- Add metrics for config reloads
  - Adds gauges to help debug issues with config reloads
- Add metric reporting config hash
  - Adds a gauge reporting a sha256 hash of the config (before
    expansion) via label. This should be useful to debug config reload
    issues.
  - The gauge is implemented for agent, agentctl and agent_flow and
    supports static, dynamic and remote config.
 
#### Which issue(s) this PR fixes

Fixes #2040 

#### Notes to the Reviewer

- I think I instrumented the code in the right places but might have missed something, please double check
- ~I had to namespace the metric name for flow as I would get a double registration otherwise. Amy ideas on how to do this better without too much complexity?~
- Is there a risk of cardinality explosion for the hash metric? (for example with dynamic configs)
- I assume we don't test internal metrics
- I assume we don't document these metrics, please correct if I am wrong
- Anything else I missed?

#### PR Checklist

- [x]  CHANGELOG updated

mattdurham commented on cmd/agent/entrypoint.go:
```diff
@@ -61,6 +65,14 @@ func NewEntrypoint(logger *server.Logger, cfg *config.Config, reloader Reloader)
+				Name: "agent_config_last_reload_successful",
+				Help: "Config loaded successfully.",
+			}),
+			configReloadSeconds: promauto.NewGauge(prometheus.GaugeOpts{
```
Check with Goutham but I wonder if its better to have success/failure be a label so we can track the two separately, it does break some standing with prometheus but might be acceptable.  So would become `agent_config_last_reload_timestamp_seconds` with label like `result=success` or `result=failure`.

jcreixell replied:
I thought about, it's probably a good idea, let's do it

mattdurham commented on cmd/agent/entrypoint.go:
```diff
@@ -250,17 +262,21 @@ func (ep *Entrypoint) TriggerReload() bool {
 
 	cfg, err := ep.reloader()
 	if err != nil {
+		ep.configReloadSuccess.Set(0)
```
Also in thinking about this I wonder if we need to change it to just `load`, or set values to good on startup. If I am alerting based on these they will not be valid til a reload happens.

gouthamve replied:
Shouldn't the default be success? This is because the "first" load of the config is a success and if its bad config, the startup itself will fail. I would default to success tbh.

jcreixell replied:
uhm, let's move this to config.go so that we keep track of loads and not just reloads

mattdurham commented on pkg/config/config.go:
```diff
@@ -434,3 +452,11 @@ func CheckSecret(t *testing.T, rawCfg string, originalValue string) {
 }
+
+// Create a sha256 hash of the config before expansion and expose it via
+// the agent_config_hash metric
```
nit: Robert likes comments to end with a .

mattdurham commented on cmd/agent/entrypoint.go:
```diff
@@ -61,6 +65,14 @@ func NewEntrypoint(logger *server.Logger, cfg *config.Config, reloader Reloader)
 		ep = &Entrypoint{
 			log:      logger,
 			reloader: reloader,
+			configReloadSuccess: promauto.NewGauge(prometheus.GaugeOpts{
```
Do these show up under the flow system? We should probably have them there to.

mattdurham reviewed (changes requested):
Looks solid with some feedback requested.

jcreixell commented on cmd/agent/flow_run.go:
```diff
@@ -25,6 +27,24 @@ import (
+		[]string{"sha256"},
+	)
+	configReloadSuccess = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "agent_flow_config_last_reload_successful",
```
still not super happy about this one, as both flow and non-flow metrics show up in the metrics endpoint, instead of being registered lazily. Ideas welcome

tpaschalis replied:
Do you mean that Flow metrics would show up even in the 'static' Agent mode?

If so, it might be a little annoying but not too much. We _could_ stuff everything into the Run method, but I'm not sure the readability is a good tradeoff.

mattdurham replied:
For Flow side: In `flow_run.go` instead of using promauto you could register it with the DefaultRegister. Same with static agent , instead of promauto register with the reg in `entrypoint.go`

jcreixell commented:
@mattdurham I have addressed your comments, please have another look when you have time. The only remaining issue I can think of is that both flow and non-flow metrics are exposed, however I am not sure that fixing this is worth the hassle. Thank you!

tpaschalis reviewed (approved):
LGTM! Nothing to add personally, simple and straightforward.

jcreixell commented:
@mattdurham See my last commit for a refactoring moving the duplicated code into a new package. 

I tried the simple solution of manually registering the metrics during execution, but it was quite intrusive, especially in config.go (since Load can be called multiple times, I would need to register the metrics outside that package and pass them as a parameter). In the end, I opted for using a singleton instance for the metrics, reused both in flow and regular agent.

Location of this new package and naming are up for discussion.

mattdurham commented on pkg/config/instrumentation/config_metrics.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// Expose metrics for reload success / failures.
+func (c *configMetrics) InstrumentLoad(isError bool) {
```
The `isError` through me off a bit; I was expecting more success bool.

jcreixell replied:
sounds good, let's change it

mattdurham commented on cmd/agent/flow_run.go:
```diff
@@ -108,11 +109,15 @@ func (fr *flowRun) Run(configFile string) error {
 	reload := func() error {
 		flowCfg, err := loadFlowFile(configFile)
 		if err != nil {
+			instrumentation.ConfigMetrics.InstrumentLoad(true)
```
Could this be simplified to one `defer instrumentation.ConfigMetrics.InstrumentLoad(err != nil)`?

jcreixell commented on pkg/config/instrumentation/config_metrics.go:
```diff
@@ -0,0 +1,59 @@
+		[]string{"sha256"},
+	)
+	m.configReloadSuccess = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "agent_config_last_reload_successful",
```
question: shall we be using the term `load` instead of `reload` here? Reload could be confusing given that we instrument every load, including the initial one

mattdurham replied:
Sounds reasonable.

mattdurham reviewed (approved):
Looks good to me, clean and simple

rfratto commented on pkg/config/instrumentation/config_metrics.go:
```diff
@@ -0,0 +1,59 @@
+		Name: "agent_config_last_load_successful",
+		Help: "Config loaded successfully.",
+	})
+	m.configLoadSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "agent_config_last_load_timestamp_seconds",
+		Help: "Timestamp of the last configuration load by result.",
+	}, []string{"result"})
```
Hm, one of my concerns with the `result` label is that every instance of the label will continue to be reported for the process lifetime, which might be confusing and hard to query for exactly what you're looking for. 

I think there's two choices here:

1. Remove the result label 
2. Delete all the labels on the vector every time we instrument a new load, so the /metrics endpoint only reports the metric and label corresponding to the most recent load.  

Personally, I'd go for the first; a user can still know when unsuccessful loads happened by querying for `agent_config_last_load_timestamp_seconds and agent_config_last_load_successful == 0`.

rfratto replied:
cc @mattdurham since I'm suggesting that the `result` label might cause issues

mattdurham replied:
Seems reasonable.

rfratto replied:
SGTM 👍

rfratto commented on pkg/config/instrumentation/config_metrics.go:
```diff
@@ -0,0 +1,59 @@
+	c.configHash.WithLabelValues(fmt.Sprintf("%x", hash)).Set(1)
+}
+
+// Expose metrics for load success / failures.
+func (c *configMetrics) InstrumentLoad(success bool) {
```
nit: Passing success to a function like this feels a little weird to me. You could implement this using some middleware-y function:

```go
cm.InstrumentLoad(bb, func(bb []byte) error {
  // do the load 
  // return if there was an error 
})
```

which would set all the appropriate metrics for you. 

Honestly, I'm not sure if that's really worth implementing right now. This is mostly just a nit, it feels a little strange but I wouldn't say it's a blocker.

jcreixell replied:
I agree with your point here, however looking at the config code, it is hard to agree on a single API that would work for all current implementations (including dynamic config). I am also purposely trying to separate config hash instrumentation from config load (under the assumption that it is better to build the hash before applying dynamic transformations like variable substitution, etc to prevent noise, while it is better to instrument loads as late as possible, to cover all possible failure modes). I am also open to have these assumptions challenged :)

rfratto replied:
I don't feel too strongly about it, it's something we can always change later and it's not intended to be a public facing API. It's more than fine for now 👍

rfratto reviewed (approved):
LGTM. I have some concerns about the result label for `agent_config_last_load_timestamp_seconds`, but how you solve it is ultimately up to you and you don't need a second approval from me. 

Let @grafana/grafana-agent-maintainers know when you're ready for it to be merged and someone can put it in main :)

jcreixell commented:
@grafana/grafana-agent-maintainers I think I am done here, feel free to merge if you are ok with my last commits and comment
